### ISSUE TITLE
[useHover] Timeouts in onMouseMove and onMouseEnter not cleared on unmount

### ISSUE THREAD

**Issue by ThenTech** [floating-ui/floating-ui#1705]
# Problem

I am running in kind of the same problem as #1494, whereas I get the `Warning: Can't perform a React state update on an unmounted component.` message in the console.

I have a table with quick-filters that have a remove button. This button uses `useFloating`  to provide a tooltip. **When clicking the button, the component gets unmounted, but the `onOpenChange` method still gets called, causing React to complain.**

I am using a setup with `useHover` [similar to this sandbox](https://codesandbox.io/s/frosty-framework-sx6gkv?file=/src/App.js). Although I cannot reproduce it there, maybe because React doesn't trigger that warning on `codesandbox`?

Custom `onOpenChange` method I am using:

```javascript
  const onOpenChange = useCallback(
    (open) => {
      setOpen(open); // Warning trigger
      if (open) {
        setCurrentId(uuid);
      }
    },
    [uuid, setCurrentId]
  );
```

# Steps to reproduce locally

(set `delay.open` to a larger value to make it easier)

## onMouseMove

[`useHover.onMouseMove`](https://github.com/floating-ui/floating-ui/blob/63581afcb8c94ac5e963a928f58ac73c083430b6/packages/react-dom-interactions/src/hooks/useHover.ts#L280)

```javascript
      onMouseMove() {
        if (open || restMs === 0) {
          return;
        }

        clearTimeout(restTimeoutRef.current);
        restTimeoutRef.current = setTimeout(() => {
          if (!blockMouseMoveRef.current) {
            onOpenChange(true);
          }
        }, restMs);
      },
    },
```

1. Hover over the reference while moving the mouse, and keep it on the target (tooltip won't show until open delay)
2. `onMouseMove` is triggered, which starts a `setTimeout` into `restTimeoutRef`
3. Now click the target, which removes the component from the DOM
4. The `restTimeoutRef` expires and `onOpenChange(true);` is called, which in turn will attempt to update the open state on a now unmounted component.

According to the (Chrome) debugging tools, both `refs.floating` and `refs.reference` are already `null` at this point.

## onMouseEnter

[`useHover.onMouseEnter`](https://github.com/floating-ui/floating-ui/blob/63581afcb8c94ac5e963a928f58ac73c083430b6/packages/react-dom-interactions/src/hooks/useHover.ts#L157)

```javascript
    function onMouseEnter(event: MouseEvent) {
      clearTimeout(timeoutRef.current);
     
      // ...

      if (openDelay) {
        timeoutRef.current = setTimeout(() => {
          onOpenChangeRef.current(true);
        }, openDelay);
      } else {
        onOpenChangeRef.current(true);
      }
    }
```

1. Start hovering over the reference (tooltip won't show until open delay)
2. `onMouseEnter` is triggered, which starts a `setTimeout` into `timeoutRef`
3. Click the target, which removes the component from the DOM
4. The `timeoutRef` expires and `onOpenChangeRef.current(true);` is called, which in turn will attempt to update the open state on a now unmounted component.

According to the (Chrome) debugging tools, both `refs.floating` and `refs.reference` are already `null` at this point.

# Versions

```
@floating-ui/react-dom-interactions@^0.6.0
@floating-ui/react-dom@^0.7.0
@floating-ui/dom@^0.5.0
@floating-ui/core@^0.7.0
```

# TL;DR

Even though the event listeners are cleaned-up, if a timeout was started when the reference exists and expires after it is unmounted, the `onOpenChange` callback is triggered, potentially causing unwanted behaviour on the now unmounted component.

So a potential fix is make sure both timeouts are actually cleared on unmount. There is the `useEffect` with the [`onDismiss`](https://github.com/floating-ui/floating-ui/blob/63581afcb8c94ac5e963a928f58ac73c083430b6/packages/react-dom-interactions/src/hooks/useHover.ts#L87) function in it, which seems to be intended for this purpose. In fact, the `events.off('dismiss', onDismiss);` is actually called _before_ the timeouts expire, unlinking the event, but this doesn't actually call `onDismiss` to clear the timeouts, so maybe it should?

**Comment by atomiks** [comment]
I think React 18 removed these warnings because they aren't harmful but we should use an unmount effect to clean them up for earlier versions at least.
